### PR TITLE
Currency: adds more query support

### LINKED ISSUE #3254: Currency: Update triggering patterns & aliases
## Problem

The currency IA is undertriggering. We need to open up the patterns to allow broader triggering. The new patterns we need to support are:

```
<currency_name> exchange rate
<currency_name> conversion rate
<currency_name> price
<currency_name> to <currency_name> exchange rate
<currency_name> to <currency_name> conversion
<currency_name> to <currency_name> conversion rate
<currency_name> to <currency_name> converter
<currency_name> to <currency_name> price
```

Further, we need to update the aliases.

**mex** --> `mex peso` and `mexican peso`
**cny**  --> `chinese yuan` and `yuan`
**gbp** --> `uk`

## Solution

1. Add the following test queries to the [Currency.t](https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/t/Currency.t).

- 140 chinese yuan to u.s. dollars
- us to mexican peso
- mex peso to dollar
- yuan exchange rate
- cny to usd exchange rate
- euro to usd conversion
- yuan conversion rate

2. Add the aliases mentioned in the above problem statement to [currencyNames.txt](https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/share/spice/currency/currencyNames.txt)

3. Update the triggering in [Currency.pm](https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/lib/DDG/Spice/Currency.pm)

4. Send a Pull Request

## People to notify

@pjhampton

## Get Started
- [ ] 1) Claim this issue by commenting below
- [ ] 2) Review our [Contributing Guide](https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/CONTRIBUTING.md)
- [ ] 3) [Set up your development environment](https://docs.duckduckhack.com/welcome/setup-dev-environment.html), and fork this repository
- [ ] 4) Create a Pull Request

## Resources
- Join [DuckDuckHack Slack](https://quackslack.herokuapp.com/) to ask questions
- Join the [DuckDuckHack Forum](https://forum.duckduckhack.com/) to discuss project planning and Instant Answer metrics
- Read the [DuckDuckHack Documentation](https://docs.duckduckhack.com/) for technical help

---

Instant Answer Page: https://duck.co/ia/view/calculator

### PR DESCRIPTION
**Pull request by adityatandon007**
**Currency Conversion IA**: This PR adds support regarding currency conversion triggering

## Description of new Instant Answer, or changes
Adds support for Currency IA

## Related Issues and Discussions
Fixes #3254 

## People to notify
@moollaza @pjhampton 

---

Instant Answer Page: https://duck.co/ia/view/currency

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by pjhampton** [comment]
Thanks @adityatandon007. Can you please follow the DDH convention for naming your PRs in the future?

@moollaza please deploy to beta 👍

**Comment by moollaza** [comment]
Deloying to Beta 👍

**Comment by moollaza** [comment]
@adityatandon007 let's add support for words like "currency converter", "currency conversion", "calculator", "calculate", etc.

E.g.
- "usd to cad currency (conversion|converter)"
- "calculate 1 cad to usd"
- "4 cad to use currency calculator"

You can use this:
```perl
my $keyword_qr = qr/(?:(?<currencyKeyword>(currency ?)?(?:value|price|conversion|converter|calculator|(exchange|conversion)\srate|(value|price)\s?of)?)\s?)/i;
```

We also don't need the optional space at the beginning as we allow for that in the `guard`

Please add tests for those as well 👍

**Comment by adityatandon007** [comment]
@moollaza we need that space we are not allowing it in the `guard` 
[]( zeroclickinfo-spice/lib/DDG/Spice/Currency.pm#L38)

**Comment by adityatandon007** [comment]
@moollaza for now I have removed the space from the keyword_qr and added that in the guard before the last $keyword_qr and support for calculator has been added in the keyword_qr regex.

**Comment by moollaza** [comment]
This is looking good @adityatandon007!

> so, I tried my regex though it doesn't looks good (more repetitive), but it doesn't fail earlier cases. Let me know if you have a workaround for those cases or if you think mine version of regex is fine.

I think this works...

```perl
qr/(?:(?<currencyKeyword>(?:((currency|value|price)( (conver(sion|ter)|calculator))?|(conver(sion|ter)|calculator)|(exchange|conversion) rate)|(value|price) of) ?)/i;
```

> One more thing this PR triggers for queries like usd conversion, usd currency. you were saying we should not trigger for those cases. Should I adjust the query so it doesn't trigger for those cases.

Those queries _**should**_ trigger it. They're perfectly valid. Google shows their converter for them as well 👍

**Comment by adityatandon007** [comment]
@moolllaza thanks for the tip but you left an `)` so I got an unmatched error but I resolved the error and your query version is working fine. 

One more thing can you please deploy this to beta.

**Comment by moollaza** [comment]
@adityatandon007 this is looking good! I just merged a Currency PR though so there's a conflict here. Can you please resolve? Then we can deploy to Beta 👍

**Comment by adityatandon007** [comment]
@moollaza I tried resolving the conflict but I am getting errors while testing the file after resolving. These are the failing test: 

```#   Failed test 'Testing query usd conversion'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 53.
#     Structures begin differing at:
#          $got->{call} = '/js/spice/currency/1/usd/eur'
#     $expected->{call} = '/js/spice/currency/1/usd/usd'

#   Failed test 'Testing query convert currency 5k usd to inr'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 53.
#     Structures begin differing at:
#          $got->{call} = '/js/spice/currency/1/usd/eur'
#     $expected->{call} = '/js/spice/currency/5000/usd/inr'

#   Failed test 'Testing query usd to cad currency converter'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 53.
#     Structures begin differing at:
#          $got->{call} = '/js/spice/currency/1/usd/eur'
#     $expected->{call} = '/js/spice/currency/1/usd/cad'

#   Failed test 'Testing query usd to cad currency conversion'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 53.
#     Structures begin differing at:
#          $got->{call} = '/js/spice/currency/1/usd/eur'
#     $expected->{call} = '/js/spice/currency/1/usd/cad'

#   Failed test 'Testing query 4 cad to usd currency calculator'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 53.
#     Structures begin differing at:
#          $got->{call} = '/js/spice/currency/1/usd/eur'
#     $expected->{call} = '/js/spice/currency/4/cad/usd'
# Looks like you failed 10 tests of 120.
```
I tried to find the reason and came to this conclusion that somebody changed the results of the earlier queries and it is now defaulting to `usd` to `eur` for those test cases. This is happening because of this piece of code 
```handle query_lc => sub {

    # returns for plain language queries such as 'currency converter'
    if(m/$lang_qr/) {
        my $from = getLocalCurrency();
        my $to = 'usd';

        if($from eq $to) {
            $to = 'eur';
        }

        return checkCurrencyCode(1, $from, $to);
    }
    ```
I don't know why we are defaulting to these. If I remove this piece of code it works fine, ie falls back to our earlier responses. One more thing some of our test cases were deleted too from test files, I don't know what to do? Should I simply go and replace my changes with the new ones (to just get rid of the conflict), if it is so then the aim which this PR is supposed to do (query support) would be simply lost. 

P.S. I want your suggestions on these.

**Comment by pjhampton** [comment]
It seems `if(m/$lang_qr/)` is not strict enough. You are going to have to tighten that regex so it has a full match, rather than a partial match. This was an oversight on my behalf but we can't merge this until it's fixed.

**Comment by adityatandon007** [comment]
@pjhampton If I use your `$keyword_qr` and `$guard` of course with `$lang_qr` it is failing these `17`  test: 
```
#   Failed test 'Expected result but dont get one on us dollar conversion rate'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on bitcoin converter'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on value of inr'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on price of bitcoin'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on usd conversion'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on usd to inr exchange rate'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on inr to cny conversion'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on 500 cny to usd conversion rate'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on 5k usd to yuan converter'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on convert currency 5k usd to inr'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.
#   Failed test 'Expected result but dont get one on conversion 5k usd to yuan'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on conversion 5k usd to uk'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on usd to cad currency converter'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on usd to cad currency conversion'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on 4 cad to usd currency calculator'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on usd to cad calculator'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.
# Looks like you failed 17 tests of 120.
```
Any suggestions on how to resolve the conflict would be helpful?

**Comment by pjhampton** [comment]
Are these your new tests?

File: share/spice/currency/currencyNames.txt [hunk 1]
```diff
@@ -44,7 +44,7 @@ etb,ethiopian birr,ethiopia birr,
 eur,euro,euros,
 fjd,fijian dollar,fiji dollar,
 fkp,falkland island pound,
-gbp,british pound,united kingdom pound,uk pound,uk pounds,pound,sterling,pound sterling,
+gbp,british pound,united kingdom pound,uk pound,uk pounds,pound,sterling,pound sterling,uk
```
Comment (pjhampton) [inline share/spice/currency/currencyNames.txt]: Let's remove `uk`. This makes no sense

**Comment by adityatandon007** [comment]
@pjhampton Please review

**Comment by pjhampton** [comment]
No problem. Thanks for this @adityatandon007! I will review today. 
Is there anything else you need from me to get this past the finish line?

**Comment by adityatandon007** [comment]
@pjhampton I am finished with this. I think only @moollaza has the right to deploy it to beta. Can you deploy it to beta as well. I w'd love to see this live. It would be very kind of you it gets deployed today?

**Comment by pjhampton** [comment]
@adityatandon007 @moollaza is a while for a while. He should be back soon. 
I will see what I can do as I want this pushed through too :+1:

File: t/Currency.t [hunk 2]
```diff
@@ -425,6 +425,12 @@ ddg_spice_test(
         caller => 'DDG::Spice::Currency',
         is_cached => 0
     ),
+				'currency calculator' => test_spice(
```
Comment (pjhampton) [inline t/Currency.t]: oops... let's move this back

Comment (pjhampton) [inline t/Currency.t]: I can sort in my pr tomorrow 👍

**Comment by pjhampton** [comment]
For anybody's reference. We hit this with the internal query tool and all was working great. Just a minor formatting issue which I will fix up tomorrow.

Thanks for your hard work as usual @adityatandon007 💯  🏅 ✨

**Comment by bsstoner** [comment]
This was released to production 🚀

**Pull request merged**